eth/vm/state.py

```diff
@@ -99,9 +103,19 @@ def logger(self) -> ExtendedDebugLogger:
         return cast(ExtendedDebugLogger, normal_logger)
 
     #
-    # Block Object Properties (in opcodes)
+    # Tracing
     #
+    @contextlib.contextmanager
+    def trace(self, tracer: BaseTracer) -> Iterator[None]:
```

Comment:
I think the idea behind this context manager is that `BaseState` instances will outlive transactions, so tracer should only be set while it's running a transaction. But I think that's really a sign that tracer shouldn't be an attribute of `BaseState` at all.

Using a context manager for this seems error-prone and unnecessary. It's error-prone because you have to remember to use this whenever you're calling a method which might need the tracer. If tracer was passed in as a parameter then the mistake of not passing one would be much harder to make.

It's unnecessary because this code seems to take two approaches simultaneously. It both adds the tracer to vm_state, making it available nearly everywhere, and it also manually threads tracer through all the calls. Because the threading has already been done, it's simpler to use that tracer everywhere.